Ready: Created the HTTP API classes and written a first endpoint to fetch information about your channel

devos50 opened the pull request:
In this PR, I created the classes that are setting up the Tribler HTTP API. The API can be enabled or disabled by using the session configuration file. Also, changing the port is possible. By default, the API listens on port 8085 (we should fix this port so the new Qt GUI knows which port to use) but it's also possible to assign a random port to the API as I do in the unit tests.

The first endpoint of the API is to fetch the overview of your channel. This returns the name, identifier (hex-encoded) and description of your channel. The API will be gradually improved and expanded in separate PRs after this one has been merged. I will first focus on the endpoints related to your channel (getting/adding torrents, playlists and rss-feeds) as requested by @brussee.

I've also written a document describing the API and available endpoints in more detail. This document might grow over time so at one point, we might have to split it into separate files. The structure of the API documentations resembles the [Twitch API documentation](https://github.com/justintv/Twitch-API) which I think has a clear structure.

I've added docstrings where I think they might increase the understandability of the code.

Also, this PR got a little larger than I initially hoped but I think it's very hard to split this since everything I did is necessary to create the base for the API.

devos50 commented:
retest this please

devos50 commented:
@lfdversluis can you review this PR please?

devos50 commented:
Hmm I was a little too quick with this PR and I was thinking about two minor modifications:
-the response type of the request should be text/json
-I want to add the configuration type of the channel to the response (open/semi-open/closed), this is displayed on the settings page of your channel

I will implement this tomorrow morning

devos50 commented:
The mac slave went offline again during the build

devos50 commented:
retest this please

devos50 commented:
I only added `text/json` as response content header. It turns out that getting the channel state (open/semi-open/closed) is more complicated than I thought and requires a Dispersy instance so I will leave this for a separate PR. @lfdversluis can you review?

lfdversluis commented on Tribler/Core/Modules/api/api_manager.py:
```diff
@@ -0,0 +1,25 @@
+        self.session = session
+        self.site = None
+
+    def start(self):
```
doc missing

lfdversluis commented on Tribler/Core/Modules/api/api_manager.py:
```diff
@@ -0,0 +1,25 @@
+        self.site = reactor.listenTCP(self.session.get_http_api_port(),
+                                      server.Site(RootEndpoint(self.session)))
+
+    def stop(self):
```
doc missing

lfdversluis commented on Tribler/Core/Modules/api/my_channel_endpoint.py:
```diff
@@ -0,0 +1,54 @@
+        resource.Resource.__init__(self)
+        self.session = session
+
+    def getChild(self, path, request):
```
doc missing (I know it's builtin but I think it can use some doc?)

lfdversluis commented on Tribler/Core/Modules/api/my_channel_endpoint.py:
```diff
@@ -0,0 +1,54 @@
+        self.session = session
+        self.channel_db_handler = self.session.open_dbhandler(NTFY_CHANNELCAST)
+
+    def return_404(self, request):
```
doc missing

lfdversluis commented on Tribler/Core/Modules/api/my_channel_endpoint.py:
```diff
@@ -0,0 +1,54 @@
+        request.setResponseCode(404)
+        request.finish()
+
+    def render_GET(self, request):
```
doc missing

lfdversluis commented on Tribler/Core/SessionConfig.py:
```diff
@@ -730,6 +730,37 @@ def get_watch_folder_path(self):
+        Sets the HTTP API listen port.
+        :param http_api_port: An integer, indicating the port where the HTTP API should listen on.
+        """
+        return self.sessconfig.set(u'http_api', u'port', http_api_port)
```
Hmmm maybe make getRandomPort so that even though this port is not in use (keep a global list or something and add this to it), it will not use this port if the http_api_port is enabled?

devos50 replied:
I'm not really sure what you mean by this. Can you explain your idea in more detail?

lfdversluis replied:
getRandomPort fetches generates a random number and checks if this port is in use. Suppose now that your http api port is enabled, so you are going to use e.g. port 8081. Now getRandom port checks by chance 8081 and sees that it is not in use, so it will use that one. The moment you want to use 8081 for your http you get an exception.

My idea is to create a list of numbers (socks server, http api port, libtorrent port, etc.) that may be used i.e. they are enabled in the config. getRandomport is then not allowed to take any of these numbers as they may be used.

devos50 replied:
Ah, so we make some port numbers reserved. That sounds like a good idea, especially if we are going to use more hardcoded ports. I would create a separate issue for this.

lfdversluis commented on Tribler/Test/Core/Modules/Api/base_api_test.py:
```diff
@@ -0,0 +1,47 @@
+
+    def parse_response(self, response):
+        self.assertEqual(response.code, self.expected_response_code)
+        if response.code == 200:
```
No need for this right? If the assert fails the code below will never be executed. This the if and whole else branch can be deleted.

devos50 replied:
I think it's needed since it makes only sense to continue and call `readBody` when the response code is 200.

lfdversluis replied:
self.assertEqual(response.code, self.expected_response_code) does that for you?

devos50 replied:
Sometimes I'm expecting a response code other than 404. In these cases, I don't want `readBody` to be called.

lfdversluis replied:
self.expected_response_code is always 200 right?

devos50 replied:
Not always, see for instance one of my tests (`test_my_channel_overview_endpoint_no_my_channel`)

lfdversluis replied:
Oh i see the structure now. Ok my bad

lfdversluis commented on Tribler/Test/test_as_server.py:
```diff
@@ -518,6 +520,9 @@ def startTest(self, callback, min_callback_delay=5, autoload_discovery=True,
         os.chdir(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..')))
 
+        # Override default HTTP API enabled to disable API in tests
+        sessdefaults['http_api']['enabled'] = False
```
Can't you use your set function here?

devos50 replied:
Unfortunately, I cannot manipulate the `SessionConfig` object that is used when the GUI is started in tests. That's why I'm overwriting the defaults instead.

whirm replied:
Why aren't you doing this on `setupPresession()`?

devos50 replied:
@whirm `setupPreSession` is not called when running the GUI tests.

devos50 commented:
@whirm can you review this PR?

whirm commented on Tribler/Core/defaults.py:
```diff
@@ -147,6 +148,11 @@
+# HTTP API config
+sessdefaults['http_api'] = OrderedDict()
+sessdefaults['http_api']['enabled'] = True
+sessdefaults['http_api']['port'] = 8081
```
We can't have hardcoded ports on the config file, it breaks multiple Tribler instances running on the same machine. (Used in experiments, tests and such).

devos50 replied:
Ok I will change it so it will always use a random port.

whirm commented on doc/Tribler API.md:
```diff
@@ -0,0 +1,31 @@
+# Tribler API
```
Let's call it REST API, it's what it really is and it will avoid confusion with the core API.

I think the API package should be renamed to something else too.

devos50 replied:
Ok that's maybe better indeed. Do you agree if I rename the `api` package to `restapi`?

whirm replied:
sounds good to me

devos50 commented:
@whirm see my last three fixup commits.

whirm commented on Tribler/Core/Modules/restapi/api_manager.py:
```diff
@@ -0,0 +1,31 @@
+from Tribler.dispersy.taskmanager import TaskManager
+
+
+class APIManager(TaskManager):
```
RESTManager?

whirm commented:
Looks good

devos50 commented:
@whirm I've rebased, updated the shutdown procedure of the `RESTManager` and added a small test in `test_sessionconfig.py`. If it looks ok I will squash it.

brussee commented:
@devos50: Please add your new classes to Setup.py (to make it work on Android)

whirm commented:
@devos50 @brussee is right :)

devos50 commented:
@brussee @whirm which Setup.py do you mean?

devos50 commented:
@brussee @whirm I've added the new module to `setup.py`

whirm commented:
looks good, squash!

devos50 commented:
Got an error when running tests, see #2134

devos50 commented:
retest this please

devos50 commented:
@whirm ready

whirm commented on Tribler/Core/defaults.py:
```diff
@@ -147,6 +148,11 @@
 
+# HTTP API config
+sessdefaults['http_api'] = OrderedDict()
+sessdefaults['http_api']['enabled'] = True
```
Sorry for noticing now, but I don't think we want this enabled by default yet...

devos50 commented:
@whirm ok np, I've disabled it by default (see my last two fixup commits). If ok, I will squash again.

whirm commented on Tribler/Test/test_as_server.py:
```diff
@@ -260,7 +260,6 @@ def setUpPreSession(self):
         self.config.set_videoplayer(False)
         self.config.set_enable_metadata(False)
         self.config.set_upgrader_enabled(False)
-        self.config.set_http_api_enabled(False)
```
leave these lines, we really want this to stay disabled on the tests even when the REST iface gets enabled by default in the future. (in tunnel/main.py too)

devos50 commented:
@whirm done

whirm commented:
Ok then! squash it!

devos50 commented:
@whirm done
